### PR TITLE
Storage: `Content` root storage

### LINKED ISSUE #54866: Storage: Create a content root folder
extracted from https://github.com/grafana/grafana/issues/52916

### PR DESCRIPTION
**Pull request by ArturWierzbicki**
fixes https://github.com/grafana/grafana/issues/54866

This PR introduces a new SQL storage called `Content` that serves the users as the entry point to all their entities.

```
content/
├── my-folder1/
|   ├── my-nested-folder1/
|   │   └── dashX.json
│   ├── dash1.json
│   ├── dash2.json
│   └── query1.json
├── team-folder-1/
│   └── dash2.json
└── devenv/                        <==== 🚀
    └── example.json  
```

 The `Content` storage is special, as its the only storage that allows other storages to be mounted/symlinked at its root level. 

Lets say we have a storage called "Development environment" pointing at the local filesystem with a "devenv" prefix:

```
devenv/
└── dev-dashboards/
    ├── all-panels.json
    ├── home.json
    ├── panel-geomap/
    │   └── panel-geomap.json
    └── panel-graph/
        └── graph-ng.json
```

Mounting the devenv storage would shadow the existing `devenv` folder because of the conflicting name:

```
content/
├── my-folder1/
|   ├── my-nested-folder1/
|   │   └── dashX.json
│   ├── dash1.json
│   ├── dash2.json
│   └── query1.json
├── team-folder-1/
│   └── dash2.json
└── devenv/                          <==== 🚀
    └── dev-dashboards/
        ├── all-panels.json
        ├── home.json
        ├── panel-geomap/
        │   └── panel-geomap.json
        └── panel-graph/
            └── graph-ng.json
```

The `devenv/example.json` is not deleted, its just no longer accessible via `content/devenv/example.json` as mounted storages always take priority over folders from the `Content` storage. Shadowed folders are filtered out from the `List` results using the existing `PathFilter` FS API abstraction.

image

image

### CODE REVIEW

File: pkg/services/store/config.go [hunk 1]
```diff
@@ -99,11 +99,13 @@ func (c *GlobalStorageConfig) save() error {
...
 
 type RootStorageConfig struct {
-	Type        string `json:"type"`
-	Prefix      string `json:"prefix"`
-	Name        string `json:"name"`
-	Description string `json:"description"`
-	Disabled    bool   `json:"disabled,omitempty"`
+	Type             string `json:"type"`
+	Prefix           string `json:"prefix"`
+	MountPoint       string `json:"mountPoint"`
```
Comment (ryantxu) [inline pkg/services/store/config.go:104]: for now, can we just make everything that is configured live under "content"?

Later we could look at how to make a more generic symlink system, but let's cross that bridge when we get there :)

**Comment by ryantxu** [comment]
Have not dug into reasons yet... but trying to navigate in the UI leaves empty pages:
http://localhost:3000/g/

image
image

**Comment by ArturWierzbicki** [comment]
> Have not dug into reasons yet... but trying to navigate in the UI leaves empty pages: http://localhost:3000/g/
> 
> image
> 
> image

🤦‍♂️ 
image
  fixed!

File: public/app/features/storage/storage.ts [hunk 2]
```diff
@@ -131,6 +131,10 @@ class SimpleStorage implements GrafanaStorage {
       path += '.json';
     }
 
+    if (!path.startsWith('content/')) {
```
Comment (ryantxu) [inline public/app/features/storage/storage.ts:134]: :(

we still need to upload files to "resources" that will not live under content

Comment (ryantxu) [reply public/app/features/storage/storage.ts:134]: nevermind!

**Comment by ryantxu** [comment]
Pushed some UI updates so the distinction is more clear:
image

**Review by ryantxu — APPROVED** [review]

**Comment by ArturWierzbicki** [comment]
> Pushed some UI updates so the distinction is more clear: image

nicee

**Pull request merged**